### PR TITLE
Feature/ROC-1777 Enter different service address

### PR DESCRIPTION
**Pull request by robertparkinson**

### CODE REVIEW

File: src/main/app/forms/models/serviceAddress.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      const line2 = ServiceAddress.toUpperCase(value.line2)
+      const city = ServiceAddress.toUpperCase(value.city)
+      const postcode = ServiceAddress.toUpperCase(value.postcode)
+      return new ServiceAddress(defendantsAddress, line1, line2, city, postcode)
+    }
+
+    return new ServiceAddress()
+  }
+
+  static toUpperCase (value: any): string {
```
Comment (sabahirfan) [inline src/main/app/forms/models/serviceAddress.ts]: Please move this to utility class, probably in `StringUtils`

Comment (robertparkinson) [reply src/main/app/forms/models/serviceAddress.ts]: done

File: src/main/app/drafts/models/defendant.ts [hunk 2]
```diff
@@ -2,19 +2,22 @@ import { Address } from 'forms/models/address'
...
   defendantRepresented?: DefendantRepresented = new DefendantRepresented()
+  serviceAddress?: ServiceAddress = new ServiceAddress()
 
   deserialize (input: any): Defendant {
     if (input) {
       this.address = new Address().deserialize( input.address )
       this.representative = new Representative().deserialize( input.representative )
       this.defendantDetails = new DefendantDetails().deserialize(input.defendantDetails)
       this.defendantRepresented = new DefendantRepresented().deserialize(input.defendantRepresented)
+      this.serviceAddress = new ServiceAddress().deserialize(input.serviceAddress)
```
Comment (sabahirfan) [inline src/main/app/drafts/models/defendant.ts]: Please cover this in relevant unit test.

File: src/main/app/utils/stringUtils.ts [hunk 3]
```diff
@@ -4,4 +4,8 @@ export default class StringUtils {
     return !(value && value.length > 0)
   }
 
+  static toUpperCase (value: any): string {
```
Comment (sabahirfan) [inline src/main/app/utils/stringUtils.ts]: Please add unit tests

File: src/main/app/forms/models/address.ts [hunk 4]
```diff
@@ -65,6 +62,6 @@ export class Address implements Serializable<Address> {
   }
 
   toString (): string {
```
Comment (sabahirfan) [inline src/main/app/forms/models/address.ts]: I know it's not a new method, but will be good to add unit tests for this as well.

File: package.json [hunk 5]
```diff
@@ -22,7 +22,7 @@
     "prepush": "yarn test && yarn test:routes"
   },
   "dependencies": {
-    "@hmcts/cmc-common-frontend": "^1.4.11",
+    "@hmcts/cmc-common-frontend": "^1.5.11",
```
Comment (yarpo) [inline package.json]: Shouldn't be new version 1.5.0?

Format A.B.C

If you update B, C should be zero'd?
If we update A, B and C = 0?

@geek-soft

File: src/main/app/utils/stringUtils.ts [hunk 6]
```diff
@@ -4,4 +4,8 @@ export default class StringUtils {
...
+    return value ? value.toUpperCase() : value
```
Comment (yarpo) [inline src/main/app/utils/stringUtils.ts]:
```
     return value ? value.toUpperCase() : ''
```

or you can't tell this method is always returning string (as you claim in line 7).

:)

File: src/main/features/claim/views/defendant-service-address.njk [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+{% set heading = t('Address for service') %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <form novalidate method="post">
+        {{ csrfProtection(csrf) }}
+
+        {% set errorType = form.errorFor('defendantsAddress') %}
+        <fieldset id="defendantsAddress[label]" class="form-group {% if errorType %} form-group-error {% endif %}">
```
Comment (yarpo) [inline src/main/features/claim/views/defendant-service-address.njk]: why don't you use `radioGroup` macro? Haven't you copied this macro from citizen fe?

File: src/test/features/claim/routes/defendant-service-address.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      draftStoreServiceMock.resolveSave('legalClaim')
+
+      await request(app)
+        .post(ClaimPaths.defendantServiceAddressPage.uri)
+        .set('Cookie', `${cookieName}=ABC`)
+        .send({ defendantsAddress: 'NO', line1: 'Apt 99', line2: '', city: 'London', postcode: 'E1' })
+        .expect(res => expect(res).to.be.redirect.toLocation(ClaimPaths.defendantAdditionPage.uri))
+    })
+  })
+})
```
Comment (yarpo) [inline src/test/features/claim/routes/defendant-service-address.ts]: are you sure you wanted to push yarn.lock file?

File: package.json (hunk 5 above)
Comment (robertparkinson) [reply package.json]: it should have been but i did it by mistake and changing it makes no sense at the moment. when we next make a change I will update it to 1.6.0

File: src/main/features/claim/views/defendant-service-address.njk (hunk 7 above)
Comment (robertparkinson) [reply src/main/features/claim/views/defendant-service-address.njk]: macros don't currently account for a pop out for each of the radio options and we can't get the desired effect using the macros

File: src/test/features/claim/routes/defendant-service-address.ts (hunk 8 above)
Comment (robertparkinson) [reply src/test/features/claim/routes/defendant-service-address.ts]: should it be in the gitignore then if we don't push it?

File: src/main/app/utils/stringUtils.ts (hunk 6 above)
Comment (robertparkinson) [reply src/main/app/utils/stringUtils.ts]: have updated it to expect a string for value as the parameter

File: package.json (hunk 5 above)
Comment (yarpo) [reply package.json]: ok

File: src/main/app/utils/stringUtils.ts (hunk 6 above)
Comment (yarpo) [reply src/main/app/utils/stringUtils.ts]: but if you expect to be a string it means you can use:
https://www.w3schools.com/jsref/jsref_touppercase.asp
http://www.c-sharpcorner.com/UploadFile/5089e0/string-object-method-in-typescript-part-7/

which means this method is not needed.

File: src/test/features/claim/routes/defendant-service-address.ts (hunk 8 above)
Comment (yarpo) [reply src/test/features/claim/routes/defendant-service-address.ts]: donno

File: src/main/app/utils/stringUtils.ts (hunk 6 above)
Comment (robertparkinson) [reply src/main/app/utils/stringUtils.ts]: true! hadn't even realised that. just copied it from else where ha. will remove it

File: package.json (hunk 5 above)
Comment (yarpo) [reply package.json]: actually can you please make another `npm release` for common fe and then update this dependency.

File: src/main/app/drafts/models/defendant.ts (hunk 2 above)
Comment (robertparkinson) [reply src/main/app/drafts/models/defendant.ts]: done

File: src/test/app/forms/models/serviceAddress.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    it('should have valid service address details elements', () => {
+      const address = ServiceAddress.fromObject({
+        defendantsAddress: YesNo.NO.value,
+        line1: 'line1',
+        line2: 'line2',
+        city: 'city',
+        postcode: 'postcode'
+      })
+      console.log(address)
```
Comment (sabahirfan) [inline src/test/app/forms/models/serviceAddress.ts]: please remove this log statement.

**Pull request merged**